test/raft/distributed_test.cc

```diff
@@ -31,5 +40,142 @@ requires (typename M::state_t s, typename M::input_t i) {
+    // where they manage to obtain the output and when they give up.
+    // The client allocates a channel only on its local server; other replicas
+    // of the state machine will therefore not find the ID in their instances
+    // of _output_channels so they just drop the output.
+    using channels_t = std::unordered_map<cmd_id_t, promise<typename M::output_t>>;
+    using channel_handle = typename channels_t::const_iterator;
+    channels_t _output_channels;
+
+    impure_state_machine(snapshots_t<typename M::state_t>& snapshots)
+        : _val(M::init), _snapshots(snapshots) {}
+
+    future<> apply(std::vector<raft::command_cref> cmds) override {
+        for (auto& cref : cmds) {
+            auto is = ser::as_input_stream(cref);
+            auto cmd_id = ser::deserialize(is, boost::type<cmd_id_t>{});
+            auto input = ser::deserialize(is, boost::type<typename M::input_t>{});
+            auto [new_state, output] = M::delta(std::move(_val), std::move(input));
+            _val = std::move(new_state);
+
+            auto it = _output_channels.find(cmd_id);
+            if (it != _output_channels.end()) {
+                // We are on the leader server where the client submitted the command
+                // and waits for the output. Send it to them.
+                it->second.set_value(std::move(output));
+            } else {
+                // This is not the leader on which the command was submitted,
+                // or it is but the client already gave up on us and deallocated the channel.
+                // In any case we simply drop the output.
+            }
+
+            if (_aborted) {
+                co_return;
+            }
+
+            co_await make_ready_future<>(); // maybe yield
+        }
+    }
+
+    future<raft::snapshot_id> take_snapshot() override {
+        auto id = raft::snapshot_id{utils::make_random_uuid()};
+        _snapshots[id] = _val;
+        co_return id;
+    }
+
+    void drop_snapshot(raft::snapshot_id id) override {
+        _snapshots.erase(id);
+    }
+
+    future<> load_snapshot(raft::snapshot_id id) override {
+        auto it = _snapshots.find(id);
+        assert(it != _snapshots.end()); // dunno if the snapshot can actually be missing
+        _val = it->second;
+        co_return;
+    }
+
+    future<> abort() override {
+        _aborted = true;
+        // TODO: the semantics of abort are not clear.
+        // If apply is in progress, should we wait for it to finish here? (it will finish on the next iteration over cmds)
+        co_return;
+    }
+
+    template <typename F>
+    future<typename M::output_t> with_output_channel(F&& f) {
+        promise<typename M::output_t> p;
+        auto fut = p.get_future();
+        auto cmd_id = utils::make_random_uuid();
+        auto [_, inserted] = _output_channels.emplace(cmd_id, std::move(p));
+        assert(inserted);
+
+        return std::forward<F>(f)(cmd_id, std::move(fut)).finally([cmd_id, this] () {
```

Comment:
Why do you need std::forward() here?

Reply:
Also why don't you use co-routines to their fullest? You do not need `finally()`. Use regular C++ RAII:

```
auto xxx = defer([&] { _output_channels.erase(cmd_id); };
co_return co_await f(cmd_id, std::move(fut));
```

Reply:
> Why do you need std::forward() here?

Well the function is generic w.r.t `F` and I want to implement it properly, in particular preserving the value category of `F` so if `F` has multiple `operator()()` definitions (with different ref-qualifiers) I want to ensure we call the proper one.

Consider this example:
```cpp
struct Fun {
    void operator()() & {
        std::cout << "operator() &" << std::endl;
    }

    void operator()() && {
        std::cout << "operator() &&" << std::endl;
    }
};

template <typename F>
void g(F&& f) {
    std::forward<F>(f)();
}

int main() {
    g(Fun{});
    auto f = Fun{};
    g(f);
}
```

The output is
```
operator() &&
operator() &
```
(as it should be).

But if I removed `forward`, we would have
```
operator() &
operator() &
```
which is probably not what we want.

Similarly if I did
```cpp
    std::move(f)();
```
we would have
```
operator() &&
operator() &&
```
again incorrect.

The only way to preserve the value category is to use `std::forward`.